web/pdf_history.js

```diff
@@ -13,21 +13,455 @@
  * limitations under the License.
  */
 
+import { parseQueryString, waitOnEventOrTimeout } from './ui_utils';
 import { getGlobalEventBus } from './dom_events';
 
+// Heuristic value used when force-resetting `this._blockHashChange`.
+const HASH_CHANGE_TIMEOUT = 1000; // milliseconds
+
+/**
+ * @typedef {Object} PDFHistoryOptions
+ * @property {IPDFLinkService} linkService - The navigation/linking service.
+ * @property {EventBus} eventBus - The application event bus.
+ */
+
+/**
+ * @typedef {Object} PushParameters
+ * @property {string} namedDest - (optional) The named destination. If absent,
+ *   a stringified version of `explicitDest` is used.
+ * @property {Array} explicitDest - The explicit destination array.
+ * @property {number} pageNumber - The page to which the destination points.
+*/
+
+function getCurrentHash() {
+  return document.location.hash;
+}
+
+function parseCurrentHash(linkService) {
+  let hash = unescape(document.location.hash).substring(1);
+  let params = parseQueryString(hash);
+
+  let page = params.page | 0;
+  if (!(Number.isInteger(page) && page > 0 && page <= linkService.pagesCount)) {
+    page = null;
+  }
+  return { hash, page, };
+}
+
 class PDFHistory {
+  /**
+   * @param {PDFHistoryOptions} options
+   */
   constructor({ linkService, eventBus, }) {
     this.linkService = linkService;
     this.eventBus = eventBus || getGlobalEventBus();
+
+    this.initialized = false;
+    this.initialBookmark = null;
+
+    this._boundEvents = Object.create(null);
+    this._isViewerInPresentationMode = false;
+
+    // Ensure that we don't miss a 'presentationmodechanged' event, by
+    // registering the listener immediately.
+    this.eventBus.on('presentationmodechanged', (evt) => {
+      this._isViewerInPresentationMode = evt.active || evt.switchInProgress;
+    });
+  }
+
+  /**
+   * @param {string} fingerprint - The PDF documents unique fingerprint.
+   * @param {boolean} resetHistory - (optional) Reset the browsing history.
+   */
+  initialize(fingerprint, resetHistory = false) {
+    if (!fingerprint || typeof fingerprint !== 'string') {
+      console.error(
+        'PDFHistory.initialize: The "fingerprint" must be a non-empty string.');
+      return;
+    }
+    let reInitialized = this.initialized && this.fingerprint !== fingerprint;
+    this.fingerprint = fingerprint;
+
+    if (!this.initialized) {
+      this._bindEvents();
+    }
+    let state = window.history.state;
+
+    this.initialized = true;
+    this.initialBookmark = null;
+
+    this._popStateInProgress = false;
+    this._blockHashChange = 0;
+    this._currentHash = getCurrentHash();
+
+    this._currentUid = this._uid = 0;
```

Comment:
Could you describe what this represents or what it's used for?

Reply:
It's used for two things:
First of all, to be able to tell if we're moving backwards of forwards in the history, during `_popState` event handler; see https://github.com/mozilla/pdf.js/pull/8775/commits/1f41a595995baedfd9265ad93e226ba0a5a89c0c#diff-3b4ebbe223f93c9123a754b4845844e7R407.

Second of all, it's used in the `back` and `forward` methods, to navigate through the history without accidentally leaving the viewer/PDF document. With the way that the history is implemented (in browsers), there's no way to access anything except the *current* entry (which I assume is for security reasons). Hence if we just unconditionally just called `window.history.{back, forward}`, we could easily navigate away from the document.

Since I cannot image that it'd be a good idea if the [`GoBack`/`GoForward`](https://github.com/mozilla/pdf.js/blob/f54dfc63dcb116c1f97577a786372bf4c2fd0ef3/web/pdf_link_service.js#L290-L300) named actions could navigate away from the document, those parameters help avoid that.

Is this clearer, if not please ask!